### ISSUE TITLE
Heads-up: UB due to misuse of mem::zeroed will soon lead to panic 

### ISSUE THREAD

**Issue by RalfJung** [valarauca/memoization#1]
[Here](https://github.com/valarauca/memoization/blob/58e1c2c1e24fab9e3a10c553cf0a9210c4eba684/src/lib.rs#L152), this crate causes UB by ["Producing an invalid value"](https://doc.rust-lang.org/reference/behavior-considered-undefined.html). Concretely, it creates an arbitrary  `T` with `mem::zeroed()`. In the near future the call to `mem::zeroed()` [will panic to avoid UB](https://github.com/rust-lang/rust/pull/71274), and our crater run determined that this crate will be affected. (Specifically, the test suite of this crate creates a zero-initialized `String`, which is invalid.)

Maybe a good replacement is [`MaybeUninit`](https://doc.rust-lang.org/core/mem/union.MaybeUninit.html), which tracks the possibility of uninitialized values at the type level to make sure the compiler does not make any false assumptions.

**Comment by valarauca** [comment]
Will this change also be stabilizing the ability to make references to `MaybeUnit` contained value? See: https://github.com/rust-lang/rust/issues/63568

Because without that "_just porting to `MaybeUninit`_" is insufficient, instead I'll need to write a lot of extra unsafe code around `Option<T>`. 

I am just a little shocked by this change because I thought the policy was not break `std`.

**Comment by John-Nagle** [comment]
And, sure enough, almost two years later, running the example from the documentation:

```
cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/memo`
thread 'main' panicked at 'attempted to zero-initialize type `alloc::string::String`, which is invalid', /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/memoization-0.1.0/src/lib.rs:189:48
stack backtrace:
   0: rust_begin_unwind
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
   2: core::panicking::panic
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:48:5
   3: core::mem::zeroed
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/mem/mod.rs:642:9
   4: <memoization::Memoized<I,O,Func> as core::ops::deref::DerefMut>::deref_mut
             at /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/memoization-0.1.0/src/lib.rs:189:48
   5: memo::main
             at ./src/main.rs:19:3
   6: core::ops::function::FnOnce::call_once
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
john@Nagle-LTS:~/projects/rustcode/memo$ 
```

**Comment by John-Nagle** [comment]
And the crate now fails its own tests:

```
failures:

---- test::test_memoized_2 stdout ----
thread 'test::test_memoized_2' panicked at 'attempted to zero-initialize type `alloc::string::String`, which is invalid', src/lib.rs:152:48
stack backtrace:
   0: rust_begin_unwind
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
   2: core::panicking::panic
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:48:5
   3: core::mem::zeroed
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/mem/mod.rs:642:9
   4: <memoization::Memoized<I,O,Func> as core::ops::deref::DerefMut>::deref_mut
             at ./src/lib.rs:152:48
   5: memoization::test::test_memoized_2
             at ./src/lib.rs:196:9
   6: memoization::test::test_memoized_2::{{closure}}
             at ./src/lib.rs:190:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227:5
   8: core::ops::function::FnOnce::call_once
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---- test::test_testing_pattern stdout ----
thread 'test::test_testing_pattern' panicked at 'attempted to zero-initialize type `alloc::string::String`, which is invalid', src/lib.rs:152:48
stack backtrace:
   0: rust_begin_unwind
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
   2: core::panicking::panic
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:48:5
   3: core::mem::zeroed
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/mem/mod.rs:642:9
   4: <memoization::Memoized<I,O,Func> as core::ops::deref::DerefMut>::deref_mut
             at ./src/lib.rs:152:48
   5: memoization::test::test_testing_pattern
             at ./src/lib.rs:218:9
   6: memoization::test::test_testing_pattern::{{closure}}
             at ./src/lib.rs:208:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227:5
   8: core::ops::function::FnOnce::call_once
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    test::test_memoized_2
    test::test_testing_pattern

test result: FAILED. 1 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s

error: test failed, to rerun pass '--lib'
john@Nagle-LTS:~/projects/rustcode/memoization$ 
```

**Comment by valarauca** [comment]
Thanks for the reminder @[John-Nagle](https://github.com/John-Nagle), time got away from me.

Fixed in most recent comment -> 8fe65069269ae727ac124dbb757d4718af2350f6
